pgscatalog.validate with pydantic

fyvon opened the pull request:
Completely removed the use of pandas schemas for scoring files validation and now using pydantic.
The validate package provides additional pydantic model classes on top of those implemented in the .core package.

To do:

- Additional tests
- Filename validation for formatted files
- HmPOS validation against original formatted file

fyvon committed:
Reorganised project structure to be compliant with uv

fyvon committed:
Deleted legacy .validate package directory

fyvon committed:
Rewrote readme for UV/nox

fyvon committed:
Moved error messages to constants

fyvon committed:
Forcing utf-8 encoding

fyvon committed:
Added -validate to CLI applications

fyvon committed:
Version increment

fyvon committed:
Added a method to check for validity

fyvon committed:
Improved main class doc

fyvon committed:
Added basic tests

nebfield commented on pgscatalog.utils/packages/pgscatalog.validate/poetry.toml:
poetry config file shouldn't exist 👀  (now we've moved to uv)

nebfield commented on pgscatalog.utils/packages/pgscatalog.validate/src/pgscatalog/validate/cli/validate_cli.py:
```diff
@@ -0,0 +1,226 @@
+              " and the harmonized scoring file(s) won't be performed.")
+
+
+def _run_validator(filepath: str, log_dir: str, score_dir: str, check_filename: bool, header: bool) -> None:
+    """Run the file validator"""
+    # TODO: add check_filename and score_dir support
+    file = os.path.basename(filepath)
+    filename = file.split('.')[0]
+    print(f"# Filename: {file}")
+    if log_dir:
+        log_file = f'{log_dir}/{filename}_log.txt'
+    else:
+        log_file = None
```
this all looks good but I personally prefer using [pathlib](https://docs.python.org/3/library/pathlib.html) when working with files in modern python (`os.path` is good too)

fyvon replied:
Now using `pathlib` everywhere instead of `str` file names (but leaving the support for `str` in public functions or classes)

nebfield commented on pgscatalog.utils/packages/pgscatalog.validate/src/pgscatalog/validate/lib/models.py:
```diff
@@ -0,0 +1,212 @@
+        """
+        if cls.__value_contains_non_printable_characters(v):
+            raise ValueError(errors.NON_PRINTABLE_CHAR.format(input=repr(v)))
+        if cls.__value_contains_leading_or_trailing_spaces(v):
+            raise ValueError(errors.LEADING_OR_TRAILING_SPACE.format(input=repr(v)))
```
do you prefer to throw an error here or [just trim the whitespace](https://docs.pydantic.dev/2.0/usage/types/string_types/#arguments-to-constr)?

fyvon replied:
Good point, a warning would be better here but there is no level of severity implemented at the moment. We might consider ignoring it for now and trim the whitespaces anyway during the import of the raw scores to PGS formatted scores.

fyvon replied:
Now a warning, unless "strict" mode is used.

nebfield commented on pgscatalog.utils/packages/pgscatalog.validate/src/pgscatalog/validate/lib/models.py:
```diff
@@ -0,0 +1,212 @@
+        pass
+
+    def _validate_cyp(self):
+        pass
```
is this intentionally not implemented? 😅

fyvon replied:
Added some basic validation for CYP variants

nebfield reviewed (approved):
looks great! some small comments / suggestions - please don't feel like you have to implement anything new - just tidying up the poetry config file would be good

fyvon committed:
Deleted poetry config file

 (we've moved to uv)

fyvon commented on pgscatalog.utils/packages/pgscatalog.validate/src/pgscatalog/validate/lib/validation.py:
```diff
@@ -0,0 +1,75 @@
+                return
+
+            # Validate the variants
+            index = 0
+            for line in self.parser.get_variants():
```
Agreed, will change that

fyvon commented on pgscatalog.utils/packages/pgscatalog.validate/src/pgscatalog/validate/cli/validate_cli.py:
```diff
@@ -0,0 +1,226 @@
+    file_handle = open(file, "r")
+    lines = file_handle.readlines()
+    file_handle.close()
+    return lines[-1]
```
I suppose that if we improve the point below (not using regex on a txt file) this function won't be necessary anymore

fyvon committed:
Replaced explicit index declaration with enumerate

fyvon committed:
Using pathlib over os.path and str file names

fyvon committed:
Code improvement

fyvon committed:
Moved most of args validation logic to argparse

fyvon committed:
Added support for warnings and strict mode.

Warnings are listed in a different attribute and can be ignored.
"strict" mode is a new parameter allowing to promote warnings as errors.

fyvon committed:
Improved summary report if using log files

fyvon committed:
Improved warnings display

So it's easier to filter programmatically

fyvon committed:
Added more tests

fyvon commented on pgscatalog.utils/packages/pgscatalog.validate/src/pgscatalog/validate/cli/validate_cli.py:
```diff
@@ -0,0 +1,226 @@
+            data_sum['other'].append(filename)
+    else:
+        print("!! validation process had an issue: the log file can't be found")
+        data_sum['other'].append(filename)
```
The summary is now done without re-reading the output files

fyvon committed:
Add validation for CYP variants

fyvon committed:
Returning errors with a generator to avoid OOM and adding a limit to error count

The way errors are returned by the validation main class has been significantly changed. Errors and warnings are now wrapped in the same error class.
Warnings are handled using pydantic validation context rather than using unsafe static variables.

fyvon commented:
Now returning errors with a generator to avoid OOM issues and setting a limit to the error count before stopping the validation (new parameter, default=50)
